src/synthesize.py

```diff
@@ -1,114 +1,136 @@
-
-def lazy_synthesize_all(gdfs: dict[str, GeoDataFrame]):
-    'Generates the simplest programs first, including equivalents.'
-    for gdf_name in gdfs.keys():
-        yield GDF(gdf_name)
+        Due to external implementations, ordering is pseudo-random
+    '''
+    # interleave results to attempt to maximize search space coverage
+    merge_gen = merge(gdfs)
+    sjoin_gen = sjoin(gdfs)
+    # zip stops when one of its sources ends
+    for m_candidate, s_candidate in zip(merge_gen, sjoin_gen):
+        yield m_candidate
+        yield s_candidate
+    # yield whatever remains of whichever still has values left
+    yield from sjoin_gen
+    yield from merge_gen
 
-    for gdf_name, gdf in gdfs.items():
-        for col in gdf.columns:
-            if col != 'geometry':
-                yield Dissolve(GDF(gdf_name), col)
 
-
-def make_candidate_filter(gdfs: dict[str, GeoDataFrame], target: GeoDataFrame):
+def make_candidate_filter(gdfs: GdfBindings, target: DataFrame) -> Callable[[Candidate], bool]:
     'Returns a predicate that checks if a `program` over `gdfs` matches `target`'
-    return lambda program: gdfs_equal(evaluate_program(program, gdfs), target)
+    return lambda program: GeoDataFrame.equals(program.interpret(gdfs), target)
+
 
-def lazy_synthesizer(gdfs: dict[str, GeoDataFrame], target: GeoDataFrame):
+def synth_matching(
+    synthesizer: Callable[[GdfBindings], CandidateGen], 
+    gdfs: GdfBindings, 
+    target: DataFrame,
+) -> CandidateGen:
     'An iterator over all programs over `gdfs` that match `target`'
     checker = make_candidate_filter(gdfs, target)
-    return filter(checker, lazy_synthesize_all(gdfs))
-
-def lazy_synthesize(gdfs: dict[str, GeoDataFrame], target: GeoDataFrame):
-    'Lazy counterpart to `synthesize`'
-    print(next(lazy_synthesizer(gdfs, target), 'No program found!'))
+    return filter(checker, synthesizer(gdfs))
+
+
+def lazy_synthesize(gdfs: dict[str, DataFrame], target: DataFrame) -> None | Candidate:
+    ''' Generates the minimum number of programs to get the first that 
+        matches `target` over `gdfs`
+    '''
+    if not isinstance(gdfs, GdfBindings):
+        gdfs = GdfBindings(gdfs)
+    out = next(synth_matching(program, gdfs, target), None)
+    print(out or 'No program found!')
+    return out
+
+
+def synthesize_all(gdfs: dict[str, DataFrame], target: DataFrame) -> list[Candidate]:
+    ''' Generates all programs over `gdfs` that match `target`
+
+        Warning: May be *exceedingly* slow! User beware!
+    '''
+    if not isinstance(gdfs, GdfBindings):
+        gdfs = GdfBindings(gdfs)
+    
+    print('Synthesizing all may be very slow!')
+    out = list()
+    for p in synth_matching(program, gdfs, target):
+        print('\t', p)
+        out.append(p)
+    print('Done.')
+    return out
+
+
+def binding_pairs(gdfs: GdfBindings):
+    'Returns a generator over all unique (unordered) pairs of DataFrames in gdfs'
+    return combinations(gdfs.items(), 2)
```

Comment:
I may be misunderstanding here, but do we want `combinations` or `permutations` here? IIRC, `combinations` is lexicographically ordered based on the input ordering of the iterable, whereas `permutations` is unordered. I think this would only return a single ordering for a given dictionary of `gdfs` of length 2, rather than returning both.

Reply:
You are correct! Instead of using two `yield` statements in `sjoin` and `merge` to get both orderings of each combination, we could use `permutations` instead of `combinations`. 

I wrote it this way because I thought it was conceptually easier for `binding_pairs` to produce each pair of things exactly one time, and let callers handle the ordering problem. 

At the time, I also wasn't sure if `merge` was order-sensitive but I'm pretty sure it is, may be worth following up with @ethanlee16 on that.

Reply:
Ah gotcha, sounds good!